compile.c

```diff
@@ -5789,65 +5840,112 @@ check_keyword(const NODE *node)
-          case NODE_SPLAT: {
-            NO_CHECK(COMPILE(args, "args (splat)", argn->nd_head));
-            ADD_INSN1(args, argn, splatarray, RBOOL(dup_rest));
-            if (flag) *flag |= VM_CALL_ARGS_SPLAT;
-            return INT2FIX(1);
+                int dup_rest, unsigned int *flag_ptr, struct rb_callinfo_kwarg **kwarg_ptr)
+{
+    if (!argn) return 0;
+
+    NODE *kwnode = NULL;
+
+    switch (nd_type(argn)) {
+      case NODE_LIST: {
+          // f(x, y, z)
+          int len = compile_args(iseq, args, argn, &kwnode);
+          RUBY_ASSERT(flag_ptr == NULL || (*flag_ptr & VM_CALL_ARGS_SPLAT) == 0);
+
+          if (kwnode) {
+              if (compile_keyword_arg(iseq, args, kwnode, kwarg_ptr, flag_ptr)) {
+                  len -= 1;
+              }
+              else {
+                  compile_hash(iseq, args, kwnode, TRUE, FALSE);
+              }
           }
-          case NODE_ARGSCAT:
-          case NODE_ARGSPUSH: {
-            int next_is_list = (nd_type_p(argn->nd_head, NODE_LIST));
-            VALUE argc = setup_args_core(iseq, args, argn->nd_head, 1, NULL, NULL);
-            if (nd_type_p(argn->nd_body, NODE_LIST)) {
-                /* This branch is needed to avoid "newarraykwsplat" [Bug #16442] */
-                int rest_len = compile_args(iseq, args, argn->nd_body, NULL, NULL);
-                ADD_INSN1(args, argn, newarray, INT2FIX(rest_len));
-            }
-            else {
-                NO_CHECK(COMPILE(args, "args (cat: splat)", argn->nd_body));
-            }
-            if (flag) {
-                *flag |= VM_CALL_ARGS_SPLAT;
-                /* This is a dirty hack.  It traverses the AST twice.
-                 * In a long term, it should be fixed by a redesign of keyword arguments */
-                if (check_keyword(argn->nd_body))
-                    *flag |= VM_CALL_KW_SPLAT;
-            }
-            if (nd_type_p(argn, NODE_ARGSCAT)) {
-                if (next_is_list) {
-                    ADD_INSN1(args, argn, splatarray, Qtrue);
-                    return INT2FIX(FIX2INT(argc) + 1);
-                }
-                else {
-                    ADD_INSN1(args, argn, splatarray, Qfalse);
-                    ADD_INSN(args, argn, concatarray);
-                    return argc;
-                }
-            }
-            else {
-                ADD_INSN1(args, argn, newarray, INT2FIX(1));
-                ADD_INSN(args, argn, concatarray);
-                return argc;
-            }
+
+          return len;
+      }
+      case NODE_SPLAT: {
+          // f(*a)
+          NO_CHECK(COMPILE(args, "args (splat)", argn->nd_head));
+          ADD_INSN1(args, argn, splatarray, RBOOL(dup_rest));
+          if (flag_ptr) *flag_ptr |= VM_CALL_ARGS_SPLAT;
+          RUBY_ASSERT(flag_ptr == NULL || (*flag_ptr & VM_CALL_KW_SPLAT) == 0);
+          return 1;
+      }
+      case NODE_ARGSCAT: {
+          if (flag_ptr) *flag_ptr |= VM_CALL_ARGS_SPLAT;
```

Comment:
Should we add  `// f(*a, *b) or f(a, *b) or f(*a, b, c)`  for this (maybe there are other constructions)?

Reply:
To write down them I need to check more...